NeuroID/src/main/java/com/neuroid/tracker/service/NIDSamplingService.kt

```diff
@@ -0,0 +1,101 @@
+package com.neuroid.tracker.service
+
+import com.neuroid.tracker.NeuroID
+import com.neuroid.tracker.utils.NIDLogWrapper
+import com.neuroid.tracker.utils.RandomGenerator
+import java.util.Calendar
+
+class NIDSamplingService(
+    val logger: NIDLogWrapper,
+    // we need this to be replaceable for the tests!
+    var randomGenerator: RandomGenerator
+) {
+
+    private var isSessionFlowSampled = true
+    private var lastUpdatedTime = 0L
+
+    /**
+     * given a site id, tell me if we sample events or not. the site ID will be compared
+     */
+    fun updateIsSampledStatus(siteID: String?) {
+        val currentSampleRate = retrieveSampleRate(siteID)
+        if (currentSampleRate >= MAX_SAMPLE_RATE) {
+            logger.d(msg="updateIsSampledStatus() we are sampling default")
+            isSessionFlowSampled = true
+            return
+        }
+        val randomValue = randomGenerator.getRandom(MAX_SAMPLE_RATE)
+        logger.d(msg="updateIsSampledStatus() random value: $randomValue sampleRate: $currentSampleRate")
+        if (randomValue < currentSampleRate) {
+            isSessionFlowSampled = true
+            logger.d(msg="updateIsSampledStatus() we are sampling")
+            return
+        }
+
+        logger.d(msg="updateIsSampledStatus() we are not sampling")
+        isSessionFlowSampled = false
+    }
+
+
+    /**
+     * Return the sample rate, 0 indicates sample everything, 100 indicates sample nothing!
+     */
+    private fun retrieveSampleRate(siteID: String?): Int {
+        // let updateConfig() figure out if we need to update or not
+        logger.d(msg="update config")
+        updateConfig()
+        logger.d(msg="config updated")
+
+        // is the site id passed in a linked site id in the options?, if so return the linked site
+        // id sample rate
+        if (isLinkedSiteID(siteID)) {
+            NeuroID.nidSDKConfig.linkedSiteOptions[siteID]?.let {
+                logger.d(msg="retrieveSampleRate site id is in options, ${it.sampleRate}")
+                return it.sampleRate
+            }
```

Comment:
There should no reason why this would not return a mapped value. The `isLinkedSiteID(siteID)` check will return false if there is no mapping in the options hashmap and continue to the next block below eventually reaching the bottom to return the DEFUALT_SAMPLE_RATE.